boot/freeldr/bootsect/ext.S

```diff
@@ -0,0 +1,717 @@
+
+.NotSparse:
+    // Read blocks from extent start
+    mov eax, dword ptr es:[si + EXT_EXTENT_START_OFFSET]
+    call ExtFileReadBlocks
+
+.FinalExtentsSkip:
+    // Loop to process next extent
+    loop .FinalExtentsLoop
+
+.FinalExtentsDone:
+    ret
+
+BasicReadFile:
+    push es
+    pushad
+    call ExtSetInodeSegment
+
+    // Set the correct Inode offset
+    mov si, word ptr BP_REL(ExtReadInodeIndexOffset)
+
+    // Set the old file address
+    mov eax, dword ptr BP_REL(ExtFileAddress)
+    mov dword ptr BP_REL(ExtFileAddressOld), eax
+
+    // Set the file size limit
+    mov eax, EXT_INODE_DATA_SIZE_LIMIT
+
+    // Load file size from Inode
+    mov ebx, dword ptr es:[si + EXT_INODE_SIZE_OFFSET]
+
+    // Compare and limit file size
+    cmp ebx, eax
+    jbe .BelowOrEqualSize
+    mov ebx, eax
+.BelowOrEqualSize:
+    // Store the file size in the ExtFileSize variable
+    mov dword ptr BP_REL(ExtFileSize), ebx
+
+    // Set rounded up file size
+    add ebx, dword ptr BP_REL(ExtBlockSizeInBytes)
+    dec ebx
+    mov dword ptr BP_REL(ExtFileSizeState), ebx
+
+    // Don't use the extents method if theres no extents flag
+    call ExtInodeDetectExtentsFlag
+    jz .NoExtents
+
+    // If this Inode use Extents mapping then use the extents method and skip the entire classic method
+    call BasicReadFileExtents
+    jmp .LDone
+.NoExtents:
+
+    // Set up for reading direct block addresses
+    xor ecx, ecx
+    mov cl, EXT_INODE_BLOCKS
+    add si, EXT_INODE_BLOCK_POINTER_OFFSET
+.LLoop:
+    call ExtSetInodeSegment
+
+    call ExtReadFileDone
+    jb .LDone
+
+    // Get the block address
+    mov eax, dword ptr es:[si]
+
+    // If the block address is zero, skip the block
+    test eax, eax
+    jz .LSkipBlock
+
+    // Set the file segment
+    call ExtSetFileSegment
+
+    // Read the block
+    call ExtReadBlock
+.LSkipBlock:
+    call ExtUpdateFileSize
+
+    // Increment block
+    add si, EXT_POINTER_SIZE
+
+    // Loop until all blocks are loaded
+    loop .LLoop
+.LDone:
+    popad
+    pop es
+    ret
+
+SearchFile:
+    call ExtSetOldFileSegment
+    call SwapESWithDS
+
+    xor si, si
+    mov dx, word ptr BP_REL(ExtFileSize)
+.FLoop:
+    mov eax, dword ptr ds:[si]                              // Load directory Inode
+
+    cmp si, dx                                              // End of buffer reached?
+    jae .Done                                               // Abort the search if yes
+
+    test eax, eax                                           // Check if Inode is zero
+    jz .Skip                                                // Skip this entry if yes
+
+    push si                                                 // Save SI
+    add si, EXT_DIRECTORY_ENTRY_NAME_OFFSET                 // Move to filename in entry
+    mov di, offset ExtLdrFileName                           // Load target filename address
+    mov cx, offset ExtLdrFileNameEnd - ExtLdrFileName       // Length of filename to compare
```

Comment:
Does this require the "offset" keyword only for the first label (ExtLdrFileNameEnd), but not the second (ExtLdrFileName) ?